sbot/suggest.js

```diff
@@ -0,0 +1,173 @@
+          if (states[key] === true) {
+            following.add(key)
+            updateQueue.add(key)
+          } else {
+            following.delete(key)
+          }
+        })
+      })
+    )
+  }, 5)
+
+  pull(
+    ssb.backlinks.read({
+      live: true,
+      old: false,
+      query: [{ $filter: {
+        dest: { $prefix: '@' },
+        value: { content: { type: 'about' } }
+      } }]
+    }),
+    pull.filter(msg => {
+      return msg.value && msg.value.content && ref.isFeedId(msg.value.content.about) && (typeof msg.value.content.name === 'string' || msg.value.content.image)
+    }),
+    pull.drain(msg => {
+      updateQueue.add(msg.value.content.about)
+    })
+  )
+
+  function updateLoop () {
+    if (updateQueue.size) {
+      var ids = Array.from(updateQueue)
+      updateQueue.clear()
+      update(ids, () => {
+        if (updateQueue.size) {
+          updateLoop()
+        } else {
+          setTimeout(updateLoop, 10e3)
+        }
+      })
+    } else {
+      setTimeout(updateLoop, 10e3)
+    }
+  }
+
+  function watchRecent () {
+    pull(
+      pullCat([
+        ssb.createLogStream({ reverse: true, limit: 100 }),
+        ssb.createLogStream({ old: false })
+      ]),
+      pull.drain(msg => {
+        if (!suggestCache[msg.value.author]) {
+          updateQueue.add(msg.value.author)
+        }
+
+        // update recent friends
+        if (following.has(msg.value.author)) {
+          var index = recentFriends.indexOf(msg.value.author)
+          if (~index) {
+            recentFriends.splice(index, 1)
+          }
+          recentFriends.push(msg.value.author)
+        }
+      })
+    )
+  }
+
+  function update (ids, cb) {
+    if (Array.isArray(ids) && ids.length) {
+      pull(
+        pull.values(ids),
+        Paramap((id, cb) => ssb.patchwork.profile.avatar({ id }, cb), 10),
+        pull.drain(item => {
+          suggestCache[item.id] = item
+        }, cb)
+      )
+    } else {
+      cb()
+    }
+  }
+
+  return {
+    profile: function ({ text, limit, defaultIds }, cb) {
+      defaultIds = defaultIds || []
+      update(defaultIds.filter(id => !suggestCache[id]), function (err) {
+        if (err) return cb(err)
+        if (typeof text === 'string' && text.trim().length) {
+          let matches = getMatches(suggestCache, text)
+          let result = sort(matches, defaultIds, recentFriends, following)
+          if (limit) {
+            result = result.slice(0, limit)
+          }
+
+          // add following attribute
+          result = result.map(x => extend(x, { following: following.has(x.id) }))
+
+          cb(null, result)
+        } else if (defaultIds && defaultIds.length) {
+          cb(null, defaultIds.map(id => suggestCache[id]))
+        } else {
+          let ids = recentFriends.slice(-(limit || 20)).reverse()
+          let result = ids.map(id => suggestCache[id])
+          result = result.map(x => extend(x, { following: following.has(x.id) }))
```

Comment:
It looks like the `ssb.ebt.block` is magically resolved, but now I'm seeing:

```
{ message: 'Cannot read property \'id\' of undefined',
  name: 'TypeError',
  stack: 'TypeError: Cannot read property \'id\' of undefined\n    at result.map.x (/home/christianbundy/src/ssbc/patchwork/sbot/suggest.js:132:75)\n    at Array.map (<anonymous>)\n    at /home/christianbundy/src/ssbc/patchwork/sbot/suggest.js:132:27\n    at update (/home/christianbundy/src/ssbc/patchwork/sbot/suggest.js:107:7)\n    at Object.profile (/home/christianbundy/src/ssbc/patchwork/sbot/suggest.js:114:7)\n    at Object.hooked (/home/christianbundy/src/ssbc/patchwork/node_modules/ssb-server/node_modules/hoox/index.js:10:15)\n    at Object.localCall (/home/christianbundy/src/ssbc/patchwork/node_modules/ssb-server/node_modules/muxrpc/local-api.js:31:29)\n    at Object.<anonymous> (/home/christianbundy/src/ssbc/patchwork/node_modules/ssb-server/node_modules/muxrpc/local-api.js:37:22)\n    at Object.request (/home/christianbundy/src/ssbc/patchwork/node_modules/ssb-server/node_modules/muxrpc/stream.js:48:17)\n    at PacketStream._onrequest (/home/christianbundy/src/ssbc/patchwork/node_modules/ssb-server/node_modules/packet-stream/index.js:161:17)' }
```

@mixmix any chance that's resolved in ssb-suggest? :upside_down_face:

Reply:
@christianbundy I just tried switching to ssb-suggest (first commit) in https://github.com/ssbc/patchwork/tree/use-ssb-suggest and things work without errors. (gonna try to get ssb-meme in there as well before making a PR though :D )